### ISSUE TITLE
Failed to compile after replacing the closure with the function itself

### ISSUE THREAD

**Issue by Vrekia** [rust-lang/rust-clippy#13094]
### Summary

**After I replaced the closure with the function,, the program failed to compile.**

### Lint Name

redundant closure

### Reproducer

I tried this code:

```rust
fn general_matrix_mul<T>(
    mat_a: &[Vec<T>],
    mat_b: &[Vec<T>],
    add: impl Fn(T, T) -> T,
    mul: impl Fn(T, T) -> T,
) -> Vec<Vec<T>>
where
    T: Clone,
{
    // C(i,j) = Σ（0,k-1) A(i,k) * B(k,j)
    let m = mat_a.len();
    let n = mat_b.len();
    (0..m)
        .map(|i| {
            (0..n)
                .map(|j| {
                    (0..m)
                        .map(|k| mul(mat_a[i][k].clone(), mat_b[k][j].clone()))
                        .reduce(|a, b| add(a, b))
                        .expect("Matrix must be qualified!")
                })
                .collect()
        })
        .collect()
}
```

I saw this happen:

```
warning: redundant closure
  --> src\matrix.rs:33:33
   |
33 |                         .reduce(|a, b| add(a, b))
   |                                 ^^^^^^^^^^^^^^^^ help: replace the closure with the function itself: `add`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure
   = note: `#[warn(clippy::redundant_closure)]` on by default
```

After I followed Clippy's suggestion, the program wouldn't compile.

```
   |
18 |     add: impl Fn(T, T) -> T,
   |     --- captured outer variable
...
28 |         .map(|i| {
   |              --- captured by this `FnMut` closure
29 |             (0..n)
30 |                 .map(|j| {
   |                      ^^^ `add` is moved here
...
33 |                         .reduce(add)
   |                                 ---
   |                                 |
   |                                 variable moved due to use in closure
   |                                 move occurs because `add` has type `impl Fn(T, T) -> T`, which does not implement the `Copy` trait
   |
help: if `impl Fn(T, T) -> T` implemented `Clone`, you could clone the value
  --> src\matrix.rs:18:10
   |
18 |     add: impl Fn(T, T) -> T,
   |          ^^^^^^^^^^^^^^^^^^ consider constraining this type parameter with `Clone`
...
33 |                         .reduce(add)
   |                                 --- you could clone this value
   ```

### Version

```text
rustc 1.79.0 (129f3b996 2024-06-10)
binary: rustc
commit-hash: 129f3b9964af4d4a709d1383930ade12dfe7c081
commit-date: 2024-06-10
host: x86_64-pc-windows-gnu
release: 1.79.0
LLVM version: 18.1.7
```

**Comment by tesuji** [comment]
I don't know why but on nightly it should suggest using `&add` instead of `add`, which should make rustc compile just fine.
```diff
   |
33 |                         .reduce(|a, b| add(a, b))
-   |                                 ^^^^^^^^^^^^^^^^ help: replace the closure with the function itself: `add`
+   |                                 ^^^^^^^^^^^^^^^^ help: replace the closure with the function itself: `&add`
 
```

**Comment by dgramop** [comment]
I get false positives on enum-struct constructions.

```
enum Error {
   Variant(ContainedType)
}

something.map_err(|e| Variant(e))
```

to appease clippy, I put braces:
```
something.map_err(|e| { Variant(e) })
```

but `cargo fmt` removes them, angering clippy.

**Comment by eliminmax** [comment]
This also happens for `extern "C"` functions.

**Comment by samueltardieu** [comment]
Issue appears to be fixed already.
